lib/ansible/plugins/inventory/tower.py

```diff
@@ -0,0 +1,183 @@
+
+# Example for reading from environment variables:
+
+# Set environment variables:
+# export TOWER_HOST=YOUR_TOWER_HOST_ADDRESS
+# export TOWER_USERNAME=YOUR_TOWER_USERNAME
+# export TOWER_PASSWORD=YOUR_TOWER_PASSWORD
+# export TOWER_INVENTORY=THE_ID_OF_TARGETED_INVENTORY
+# Read the inventory specified in TOWER_INVENTORY from Ansible Tower, and list them.
+# The inventory path must always be @tower_inventory if you are reading all settings from environment variables.
+# ansible-inventory -i @tower_inventory --list
+'''
+
+import re
+import os
+import json
+from ansible.module_utils import six
+from ansible.module_utils.urls import Request, urllib_error, ConnectionError, socket, httplib
+from ansible.module_utils._text import to_native
+from ansible.errors import AnsibleParserError
+from ansible.plugins.inventory import BaseInventoryPlugin
+
+# Python 2/3 Compatibility
+try:
+    from urlparse import urljoin
+except ImportError:
+    from urllib.parse import urljoin
+
+
+class InventoryModule(BaseInventoryPlugin):
+    NAME = 'tower'
+    # Stays backward compatible with tower inventory script.
+    # If the user supplies '@tower_inventory' as path, the plugin will read from environment variables.
+    no_config_file_supplied = False
+
+    def read_tower_inventory(self, tower_host, tower_user, tower_pass, inventory, verify_ssl=True):
+        if not re.match('(?:http|https)://', tower_host):
+            tower_host = 'https://{tower_host}'.format(tower_host=tower_host)
+        inventory_id = inventory.replace('/', '')
+        inventory_url = '/api/v2/inventories/{inv_id}/script/?hostvars=1&towervars=1&all=1'.format(inv_id=inventory_id)
+        inventory_url = urljoin(tower_host, inventory_url)
+
+        request_handler = Request(url_username=tower_user,
+                                  url_password=tower_pass,
+                                  force_basic_auth=True,
+                                  validate_certs=verify_ssl)
+
+        try:
+            response = request_handler.get(inventory_url)
+            if response.getcode() != 200:
+                # If the GET /api/v2/inventories/N/script is not HTTP 200, print the error code
+                raise AnsibleParserError(to_native('Connection to remote host failed: {response}'
+                                                   .format(response=response)))
+            try:
+                # Attempt to parse JSON
+                return json.loads(response.read())
+            except (ValueError, TypeError) as e:
+                # If the JSON parse fails, print the ValueError
+                raise AnsibleParserError(to_native('Failed to parse json from host: {err}'.format(err=e)))
+        except (ConnectionError, urllib_error.URLError, socket.error, httplib.HTTPException) as e:
+            raise AnsibleParserError(to_native('Connection to remote host failed: {err}'.format(err=e)))
+
+    def verify_file(self, path):
+        if path.endswith('@tower_inventory'):
```

Comment:
should this really be 'endswith' or more like equal? `path.strip() == '@tower_inventory'`

Reply:
On my computer it gives `/Users/yunfzhan/ansible/@tower_inventory`, so we probably should stick with `endswith`.

Reply:
ah, its seeing it as a file name ... weird since host_list 'works' .. i should look into that

Reply:
This is likely to be the use pattern of Tower / AWX, so I think allow this may be important. It sounds like this implementation may be subject to revision if another general fix is put in, but is there anything blocking this PR @bcoca for the time being?